### DISCUSSION TITLE
Insert conversation starters generated by AI for your future conversations

### DISCUSSION THREAD

**Discussion by louis030195** [SilentVoid13/Templater#952] [Templates Showcase]
I often prepare my conversations beforehand, whether for business or family/friends, with this Templater you can get ready for great & deep conversations using https://langa.me API (free tier).

```js
const langame = async (topics) => {
    document.body.style.cursor = "wait";
    const url = `https://api.langa.me/v1/conversation/starter`;
    const langameApiKey = "grab me here https://langa.me";
    topics = topics.split(",");
    const response = await fetch(url, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            "X-Api-Key": langameApiKey,
            "accept-encoding": "*",
            "accept": "*/*"
        },
        body: JSON.stringify({
            "topics": [topics],
            "limit": 3,
        })
    }).then(response => response.json());
    document.body.style.cursor = "default";
    return response.results.map((e) => "\n- " + e.conversation_starter.en);
}
module.exports = langame;
```

```md
<% tp.user.langame(await tp.system.prompt()) %>
```

<img width="698" alt="image" src="https://user-images.githubusercontent.com/25003283/207103523-4029b487-6aa0-48dd-acac-f528c024a839.png">

<img width="657" alt="image" src="https://user-images.githubusercontent.com/25003283/207103181-54268911-59b4-4028-bfdb-94869aca124f.png">

**Comment by masonroyal** [comment]
This looks very interesting, but I'm a little confused on how to use this with templater. Could you please explain how to use this code?

**Reply by louis030195** [reply]
I don't understand? Just add the template to your hotkeys for example, I use cmd+shift+k

Before having a conversation with someone I write my thoughts in the note and generate some conversation starters using the shortcut.

The configuration is like any Templater scripts:
- a .js with the above script
- a .md with the above md code

Templater configured to detect my .js script files and my .md files


## My config

<img width="822" alt="image" src="https://user-images.githubusercontent.com/25003283/207274126-308a8ee7-c3f3-46fc-acf6-663708c6987c.png">


<img width="838" alt="image" src="https://user-images.githubusercontent.com/25003283/207274205-72326cd6-273f-4a0c-ad97-12544c2e4b93.png">

**Reply by masonroyal** [reply]
Thanks for your response! I am pretty new to scripts so I'm sure it is a me problem. I've tried to play around with it, but still can't get it to work.

When I run it as is, I get the following error in the console:
Templater Error: Template parsing error, aborting. 
 myLangameBigTalksCollection is not defined


When I changed the module.exports to langame from myLangameBigTalksCollection, I get the following error:
Templater Error: Template parsing error, aborting. 
 tp.user.langame is not a function


When I changed the const and the method being called in the prompt from langame  to myLangameBigTalksCollection, it seems to make it to the API call, but then I get no output and the following error in the console:
VM425:5          POST https://api.langa.me/v1/conversation/starter 401


I am pretty sure I am doing something wrong with my variables here. I am not sure why module.exports is assigned myLangameBigTalksCollection and where that variable is supposed to come from.

 I apologize too if my technical communication is subpar, I am only a couple of months into my javascript journey.

**Reply by louis030195** [reply]
> Thanks for your response! I am pretty new to scripts so I'm sure it is a me problem. I've tried to play around with it, but still can't get it to work.
> 
> When I run it as is, I get the following error in the console: Templater Error: Template parsing error, aborting. myLangameBigTalksCollection is not defined
> 
> When I changed the module.exports to langame from myLangameBigTalksCollection, I get the following error: Templater Error: Template parsing error, aborting. tp.user.langame is not a function
> 
> When I changed the const and the method being called in the prompt from langame to myLangameBigTalksCollection, it seems to make it to the API call, but then I get no output and the following error in the console: VM425:5 POST https://api.langa.me/v1/conversation/starter 401
> 
> I am pretty sure I am doing something wrong with my variables here. I am not sure why module.exports is assigned myLangameBigTalksCollection and where that variable is supposed to come from.
> 
> I apologize too if my technical communication is subpar, I am only a couple of months into my javascript journey.

@masonroyal 
Deeply sorry - updated the instruction code, indeed the function name is `langame` and it's exported as `langame`and used in the markdown script as `langame`also in Templater settings you will see named `langame`

401 error means "Unauthorized" https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401

So I guess you didn't set your key or wrong one, you need to go on https://langa.me and get a key here:

![image](https://user-images.githubusercontent.com/25003283/208039125-ca0ee8d6-e6c3-431f-99e6-e1d7338f478c.png)

Clicking the icon will copy the key value to your clipboard, then you can replace the string value of `langameApiKey` by it.

```js
const langameApiKey = "grab me here https://langa.me";
```

## Optional

You can store secret keys in a specific file, for example "secrets.json"

```js
// YOUR_VAULT_PATH/secrets.json
// { "langame": { "apiKey": "fooBar"}}
const conf = JSON.parse(fs.readFileSync(app.vault.adapter.basePath + '/secrets.json'))
// then const langameApiKey = conf.langame.apiKey
```

Especially to avoid publishing these secrets to the web or Git

**Reply by masonroyal** [reply]
Thank you so much! Everything is working now. Somehow I did miss the the api key section. 
I hadn't heard of langa.me before. Really excited for the possibilities of this.

And thank you for the tip on storing api keys. If anyone else is trying to make this work and having trouble, I couldn't get it to work with the .parse method, but got it work with the following:
`const conf = require(app.vault.adapter.basePath + '/secrets.json');`

**Reply by AnthonyPHN** [reply]
Hi thanks for your idea and script. I ran it but the problem is after I entered the key words there not thing coming out and the line disappeared.  Do you ever have the same problem? and what could be the fix. Thanks

**Reply by louis030195** [reply]
> Hi thanks for your idea and script. I ran it but the problem is after I entered the key words there not thing coming out and the line disappeared. Do you ever have the same problem? and what could be the fix. Thanks

@AnthonyPHN can you share your console logs, please? (option+cmd+i on mac, something similar on other hardware)

**Reply by AnthonyPHN** [reply]
![image](https://user-images.githubusercontent.com/121584940/210207436-cfae94a6-31df-4ff7-bc53-ffb2182f508f.png)
Here it is

**Reply by louis030195** [reply]
what did you write in the modal?

**Comment by circuitsnacks** [comment]
hi, this is awesome.  Thanks for creating it.   I'm attempting to set it up, but it seems that langa.me isn't hosted anymore?  Am I able to use this with my chatgpt api?

**Reply by louis030195** [reply]
langa.me is dead leftover code: https://github.com/langa-me

you can do the same with gpt4 or chatgpt api

see code here to use gpt4 https://github.com/different-ai/embedbase-docs/blob/421b1eeb6bdf74974e9d8e5f69c9f488109e6e6b/pages/api/qa.ts#L16 or https://platform.openai.com/docs/api-reference/chat/create?lang=node.js

just prompt it "give me a conversation starter for my future conversation about TOPICS, don't say anything else"
